File: crates/core/component/shielded-pool/src/gnark/shielded_ics20_withdrawal_witness_binary.rs
use anyhow::{bail, Context, Result};

use crate::{
    gnark::{
        binary::{
            encode_triple_path_32, encode_vec_32, put_bytes, put_u32, put_u64, put_u8, BinaryCursor,
        },
        shielded_ics20_withdrawal_witness::{
            ShieldedIcs20WithdrawalChangeWitness, ShieldedIcs20WithdrawalOptionalSpendWitness,
            ShieldedIcs20WithdrawalRequiredSpendWitness, ShieldedIcs20WithdrawalWitness,
        },
        transfer_witness_binary::{decode_volume_accumulator, encode_volume_accumulator},
        typed::{
            decode_indexed_leaf, encode_indexed_leaf, encode_merkle_path, encode_point_affine,
        },
    },
    ShieldedIcs20WithdrawalFamilyId,
};

const SHIELDED_ICS20_WITHDRAWAL_WITNESS_MAGIC: &[u8; 4] = b"PIWG";

impl ShieldedIcs20WithdrawalWitness {
    pub fn encode(&self) -> Result<Vec<u8>> {
        let mut buf = Vec::new();
        put_bytes(&mut buf, SHIELDED_ICS20_WITHDRAWAL_WITNESS_MAGIC);
        put_u32(&mut buf, 0);
        put_u32(&mut buf, self.family_id.get());
        put_u32(&mut buf, self.n_in);
        put_bytes(&mut buf, &self.anchor);
        put_bytes(&mut buf, &self.asset_anchor);
        put_bytes(&mut buf, &self.compliance_anchor);
        put_bytes(&mut buf, &self.target_timestamp);
        put_bytes(&mut buf, &self.outbound_asset_id);
        put_bytes(&mut buf, &self.outbound_amount);
        for limb in &self.withdrawal_effect_hash_limbs {
            put_bytes(&mut buf, limb);
        }
        put_bytes(&mut buf, &self.claimed_statement_hash);
        put_bytes(&mut buf, &self.routing_tag);
        put_bytes(&mut buf, &self.routing_parameter_set_id);
        put_bytes(&mut buf, &self.recent_position_floor);
        encode_volume_accumulator(&mut buf, &self.volume_accumulator)?;
        put_bytes(&mut buf, &self.action_balance_blinding);
        put_bytes(&mut buf, &self.nk);
        encode_merkle_path(&mut buf, &self.asset_path)?;
        put_u64(&mut buf, self.asset_position);
        encode_indexed_leaf(&mut buf, &self.asset_indexed_leaf);
        put_u8(&mut buf, u8::from(self.is_regulated));
        put_u8(&mut buf, self.regulated_precision);
        put_u8(&mut buf, self.unregulated_precision);
        put_u64(&mut buf, self.routing_as_of_height);
        put_bytes(&mut buf, &self.routing_nonce);
        encode_merkle_path(&mut buf, &self.sender_compliance_path)?;
        put_u64(&mut buf, self.sender_compliance_position);
        put_bytes(&mut buf, &self.sender_d);
        put_bytes(&mut buf, &self.sender_status);
        put_bytes(&mut buf, &self.compliance_nonce_root);
        encode_vec_32(&mut buf, &self.detection_ciphertext)?;
        put_bytes(&mut buf, &self.ring_id_hash);
        put_bytes(&mut buf, &self.policy_id_hash);
        put_bytes(&mut buf, &self.resource_hash);
        put_bytes(&mut buf, &self.permission_hash);
        put_bytes(&mut buf, &self.metadata_target_timestamp);
        put_bytes(&mut buf, &self.sender_salt);
        put_bytes(&mut buf, &self.sender_c2);
        encode_vec_32(&mut buf, &self.sender_ciphertext)?;
        put_bytes(&mut buf, &self.sender_randomizer);
        encode_required_spend(&mut buf, &self.required_spend)?;
        encode_optional_spend(&mut buf, &self.optional_spend)?;
        encode_change_output(&mut buf, &self.change_output);
        encode_point_affine(&mut buf, &self.ak_affine);
        encode_point_affine(&mut buf, &self.sender_diversified_generator_affine);
        encode_point_affine(&mut buf, &self.sender_transmission_key_affine);
        encode_point_affine(&mut buf, &self.asset_indexed_leaf_dk_pub_affine);
        encode_point_affine(&mut buf, &self.asset_indexed_leaf_ring_pk_affine);
        encode_point_affine(&mut buf, &self.sender_epk_affine);

        let total_len = u32::try_from(buf.len())
            .context("encoded shielded ICS-20 withdrawal witness exceeds u32")?;
        buf[4..8].copy_from_slice(&total_len.to_le_bytes());
        Ok(buf)
    }

    pub fn decode(bytes: &[u8]) -> Result<Self> {
        let mut cursor = BinaryCursor::new(bytes);
        if cursor.read_fixed::<4>()? != *SHIELDED_ICS20_WITHDRAWAL_WITNESS_MAGIC {
            bail!("invalid shielded ICS-20 withdrawal witness magic");
        }
        let total_length = cursor.read_u32()?;
        if total_length as usize != bytes.len() {
            bail!(
                "shielded ICS-20 withdrawal witness length mismatch: header={}, actual={}",
                total_length,
                bytes.len()
            );
        }

        let family_id = ShieldedIcs20WithdrawalFamilyId::try_from(cursor.read_u32()?)?;
        let n_in = cursor.read_u32()?;
        if n_in as usize != family_id.input_count() {
            bail!(
                "{} witness shape mismatch: got {} inputs, expected {}",
                family_id.label(),
                n_in,
                family_id.input_count()
            );
        }

        let witness = Self {
            family_id,
            total_length,
            n_in,
            anchor: cursor.read_fixed::<32>()?,
            asset_anchor: cursor.read_fixed::<32>()?,
            compliance_anchor: cursor.read_fixed::<32>()?,
            target_timestamp: cursor.read_fixed::<32>()?,
            outbound_asset_id: cursor.read_fixed::<32>()?,
            outbound_amount: cursor.read_fixed::<32>()?,
            withdrawal_effect_hash_limbs: [
                cursor.read_u128_field()?,
                cursor.read_u128_field()?,
                cursor.read_u128_field()?,
                cursor.read_u128_field()?,
            ],
            claimed_statement_hash: cursor.read_fixed::<32>()?,
            routing_tag: cursor.read_fixed::<32>()?,
            routing_parameter_set_id: cursor.read_fixed::<32>()?,
            recent_position_floor: cursor.read_fixed::<32>()?,
            volume_accumulator: decode_volume_accumulator(&mut cursor)?,
            action_balance_blinding: cursor.read_fr()?,
            nk: cursor.read_fixed::<32>()?,
            asset_path: cursor.read_merkle_path()?,
            asset_position: cursor.read_u64()?,
            asset_indexed_leaf: decode_indexed_leaf(&mut cursor)?,
            is_regulated: cursor.read_bool()?,
            regulated_precision: cursor.read_u8()?,
            unregulated_precision: cursor.read_u8()?,
            routing_as_of_height: cursor.read_u64()?,
            routing_nonce: cursor.read_fixed::<32>()?,
            sender_compliance_path: cursor.read_merkle_path()?,
            sender_compliance_position: cursor.read_u64()?,
            sender_d: cursor.read_fixed::<32>()?,
            sender_status: cursor.read_fixed::<32>()?,
            compliance_nonce_root: cursor.read_fr()?,
            detection_ciphertext: cursor.read_vec_32()?,
            ring_id_hash: cursor.read_fixed::<32>()?,
            policy_id_hash: cursor.read_fixed::<32>()?,
            resource_hash: cursor.read_fixed::<32>()?,
            permission_hash: cursor.read_fixed::<32>()?,
            metadata_target_timestamp: cursor.read_fixed::<32>()?,
            sender_salt: cursor.read_fixed::<32>()?,
            sender_c2: cursor.read_fixed::<32>()?,
            sender_ciphertext: cursor.read_vec_32()?,
            sender_randomizer: cursor.read_fr()?,
            required_spend: decode_required_spend(&mut cursor)?,
            optional_spend: decode_optional_spend(&mut cursor)?,
            change_output: decode_change_output(&mut cursor)?,
            ak_affine: cursor.read_point_affine()?,
            sender_diversified_generator_affine: cursor.read_point_affine()?,
            sender_transmission_key_affine: cursor.read_point_affine()?,
            asset_indexed_leaf_dk_pub_affine: cursor.read_point_affine()?,
            asset_indexed_leaf_ring_pk_affine: cursor.read_point_affine()?,
            sender_epk_affine: cursor.read_point_affine()?,
        };

        cursor.finish(family_id.label())?;
        Ok(witness)
    }
}

fn encode_required_spend(
    buf: &mut Vec<u8>,
    spend: &ShieldedIcs20WithdrawalRequiredSpendWitness,
) -> Result<()> {
    put_bytes(buf, &spend.nullifier);
    put_bytes(buf, &spend.spent_note_blinding);
    put_bytes(buf, &spend.spent_note_amount);
    put_u64(buf, spend.state_commitment_position);
    encode_triple_path_32(buf, &spend.state_commitment_auth_path)?;
    put_bytes(buf, &spend.spend_auth_randomizer);
    encode_point_affine(buf, &spend.rk_affine);
    put_u8(buf, u8::from(spend.history_required));
    Ok(())
}

fn decode_required_spend(
    cursor: &mut BinaryCursor<'_>,
) -> Result<ShieldedIcs20WithdrawalRequiredSpendWitness> {
    Ok(ShieldedIcs20WithdrawalRequiredSpendWitness {
        nullifier: cursor.read_fixed::<32>()?,
        spent_note_blinding: cursor.read_fixed::<32>()?,
        spent_note_amount: cursor.read_fixed::<32>()?,
        state_commitment_position: cursor.read_u64()?,
        state_commitment_auth_path: cursor.read_triple_path_32()?,
        spend_auth_randomizer: cursor.read_fr()?,
        rk_affine: cursor.read_point_affine()?,
        history_required: cursor.read_bool()?,
    })
}

fn encode_optional_spend(
    buf: &mut Vec<u8>,
    spend: &ShieldedIcs20WithdrawalOptionalSpendWitness,
) -> Result<()> {
    encode_required_spend(buf, &spend.spend)?;
    put_u8(buf, u8::from(spend.is_dummy));
    put_bytes(buf, &spend.dummy_nullifier_seed);
    Ok(())
}

fn decode_optional_spend(
    cursor: &mut BinaryCursor<'_>,
) -> Result<ShieldedIcs20WithdrawalOptionalSpendWitness> {
    Ok(ShieldedIcs20WithdrawalOptionalSpendWitness {
        spend: decode_required_spend(cursor)?,
        is_dummy: cursor.read_bool()?,
        dummy_nullifier_seed: cursor.read_fixed::<32>()?,
    })
}

fn encode_change_output(buf: &mut Vec<u8>, output: &ShieldedIcs20WithdrawalChangeWitness) {
    put_bytes(buf, &output.note_commitment);
    put_bytes(buf, &output.created_note_blinding);
    put_bytes(buf, &output.created_note_amount);
}

fn decode_change_output(
    cursor: &mut BinaryCursor<'_>,
) -> Result<ShieldedIcs20WithdrawalChangeWitness> {
    Ok(ShieldedIcs20WithdrawalChangeWitness {
        note_commitment: cursor.read_fixed::<32>()?,
        created_note_blinding: cursor.read_fixed::<32>()?,
        created_note_amount: cursor.read_fixed::<32>()?,
    })
}

File: tools/gnark/internal/abi/shielded_ics20_withdrawal_witness_binary.go
package abi

import (
	"bytes"
	"fmt"

	"github.com/mizufinance/shieldd/tools/gnark/internal/generated"
)

const (
	shieldedIcs20WithdrawalWitnessMagic   = "PIWG"
	shieldedIcs20WithdrawalWitnessVersion = 16
	maxShieldedIcs20WithdrawalInputs      = 2
	minShieldedIcs20RequiredSpendBytes    = 32*4 + 8 + 4 + 32 + 64 + 1
	minShieldedIcs20OptionalSpendBytes    = minShieldedIcs20RequiredSpendBytes + 1 + 32
	minShieldedIcs20WithdrawalChangeBytes = 32 * 13
	minShieldedIcs20WithdrawalTailBytes   = 64 * 2
)

type ShieldedIcs20WithdrawalRequiredSpendWitnessBinary struct {
	Nullifier                   [32]byte
	SpentNoteBlinding           [32]byte
	SpentNoteAmount             [32]byte
	SpentNoteRecoveryCommitment [32]byte
	StateCommitmentPosition     uint64
	StateCommitmentAuthPath     [][3][32]byte
	SpendAuthRandomizer         [32]byte
	RKAffine                    PointAffineBinary
	HistoryRequired             bool
}

type ShieldedIcs20WithdrawalOptionalSpendWitnessBinary struct {
	ShieldedIcs20WithdrawalRequiredSpendWitnessBinary
	IsDummy            bool
	DummyNullifierSeed [32]byte
}

type ShieldedIcs20WithdrawalChangeWitnessBinary struct {
	NoteCommitment      [32]byte
	RecoveryCommitment  [32]byte
	CreatedNoteBlinding [32]byte
	CreatedNoteAmount   [32]byte
	RecoveryCapsule     RecoveryCapsuleWitnessBinary
}

type ShieldedIcs20WithdrawalWitnessBinary struct {
	TotalLength uint32
	FamilyID    uint32
	NIn         uint32

	Anchor                           [32]byte
	AssetAnchor                      [32]byte
	ComplianceAnchor                 [32]byte
	TargetTimestamp                  [32]byte
	OutboundAssetID                  [32]byte
	OutboundAmount                   [32]byte
	WithdrawalEffectHashLimbs        [4][32]byte
	ClaimedStatementHash             [32]byte
	RoutingTag                       [32]byte
	RoutingParameterSetID            [32]byte
	WithdrawalEPKAffine              PointAffineBinary
	WithdrawalC2                     [32]byte
	WithdrawalKeyConfirmation        [32]byte
	WithdrawalEncryptedSenderAddress [3][32]byte
	RecentPositionFloor              [32]byte
	ActionBalanceBlinding            [32]byte
	NK                               [32]byte
	CNK                              [32]byte

	AssetPath                MerklePathBinary
	AssetPosition            uint64
	AssetIndexedLeaf         IndexedLeafBinary
	IsRegulated              bool
	RegulatedPrecision       uint8
	UnregulatedPrecision     uint8
	RoutingAsOfHeight        uint64
	RoutingNonce             [32]byte
	SenderCompliancePath     MerklePathBinary
	SenderCompliancePosition uint64
	SenderCapkAffine         PointAffineBinary
	SenderCnkCommitment      [32]byte
	SenderStatus             [32]byte
	WithdrawalSeed           [32]byte
	WithdrawalRandomizer     [32]byte

	RequiredSpend ShieldedIcs20WithdrawalRequiredSpendWitnessBinary
	OptionalSpend ShieldedIcs20WithdrawalOptionalSpendWitnessBinary
	ChangeOutput  ShieldedIcs20WithdrawalChangeWitnessBinary

	AKAffine                     PointAffineBinary
	AssetIndexedLeafDKPubAffine  PointAffineBinary
	AssetIndexedLeafRingPKAffine PointAffineBinary
	SenderDiversifiedGenerator   PointAffineBinary
}

func DecodeShieldedIcs20WithdrawalWitness(payload []byte) (*ShieldedIcs20WithdrawalWitnessBinary, generated.ShieldedIcs20WithdrawalFamilySpec, error) {
	reader := bytes.NewReader(payload)

	magic, err := readExact(reader, 4)
	if err != nil {
		return nil, generated.ShieldedIcs20WithdrawalFamilySpec{}, err
	}
	if string(magic) != shieldedIcs20WithdrawalWitnessMagic {
		return nil, generated.ShieldedIcs20WithdrawalFamilySpec{}, fmt.Errorf("invalid shielded ICS-20 withdrawal witness magic %q", string(magic))
	}
	version, err := readU32(reader)
	if err != nil {
		return nil, generated.ShieldedIcs20WithdrawalFamilySpec{}, err
	}
	if version != shieldedIcs20WithdrawalWitnessVersion {
		return nil, generated.ShieldedIcs20WithdrawalFamilySpec{}, fmt.Errorf("unsupported shielded ICS-20 withdrawal witness version %d", version)
	}
	totalLength, err := readU32(reader)
	if err != nil {
		return nil, generated.ShieldedIcs20WithdrawalFamilySpec{}, err
	}
	if totalLength != uint32(len(payload)) {
		return nil, generated.ShieldedIcs20WithdrawalFamilySpec{}, fmt.Errorf("payload length mismatch: header=%d actual=%d", totalLength, len(payload))
	}
	familyID, err := readU32(reader)
	if err != nil {
		return nil, generated.ShieldedIcs20WithdrawalFamilySpec{}, err
	}
	family, ok := generated.ShieldedIcs20WithdrawalFamilyByID(familyID)
	if !ok {
		return nil, generated.ShieldedIcs20WithdrawalFamilySpec{}, fmt.Errorf("unknown shielded ICS-20 withdrawal family id %d", familyID)
	}
	nIn, err := readU32(reader)
	if err != nil {
		return nil, generated.ShieldedIcs20WithdrawalFamilySpec{}, err
	}
	if int(nIn) != family.NIn {
		return nil, generated.ShieldedIcs20WithdrawalFamilySpec{}, fmt.Errorf("shielded ICS-20 withdrawal witness shape mismatch: got %d inputs, expected %d", nIn, family.NIn)
	}
	if nIn > maxShieldedIcs20WithdrawalInputs {
		return nil, generated.ShieldedIcs20WithdrawalFamilySpec{}, fmt.Errorf("shielded ICS-20 withdrawal witness nIn %d exceeds max %d", nIn, maxShieldedIcs20WithdrawalInputs)
	}
	minRemaining := minShieldedIcs20RequiredSpendBytes + minShieldedIcs20OptionalSpendBytes +
		minShieldedIcs20WithdrawalChangeBytes + minShieldedIcs20WithdrawalTailBytes
	if reader.Len() < minRemaining {
		return nil, generated.ShieldedIcs20WithdrawalFamilySpec{}, fmt.Errorf(
			"shielded ICS-20 withdrawal witness too short for %d spends: remaining=%d min=%d",
			nIn, reader.Len(), minRemaining,
		)
	}

	out := &ShieldedIcs20WithdrawalWitnessBinary{
		TotalLength: totalLength,
		FamilyID:    familyID,
		NIn:         nIn,
	}
	if out.Anchor, err = read32(reader); err != nil {
		return nil, family, err
	}
	if out.AssetAnchor, err = read32(reader); err != nil {
		return nil, family, err
	}
	if out.ComplianceAnchor, err = read32(reader); err != nil {
		return nil, family, err
	}
	if out.TargetTimestamp, err = read32(reader); err != nil {
		return nil, family, err
	}
	if out.OutboundAssetID, err = read32(reader); err != nil {
		return nil, family, err
	}
	if out.OutboundAmount, err = read32(reader); err != nil {
		return nil, family, err
	}
	for index := range out.WithdrawalEffectHashLimbs {
		if out.WithdrawalEffectHashLimbs[index], err = readU128Field(reader); err != nil {
			return nil, family, err
		}
	}
	if out.ClaimedStatementHash, err = read32(reader); err != nil {
		return nil, family, err
	}
	if out.RoutingTag, err = read32(reader); err != nil {
		return nil, family, err
	}
	if out.RoutingParameterSetID, err = read32(reader); err != nil {
		return nil, family, err
	}
	if out.WithdrawalEPKAffine, err = readPointAffine(reader); err != nil {
		return nil, family, err
	}
	if out.WithdrawalC2, err = read32(reader); err != nil {
		return nil, family, err
	}
	if out.WithdrawalKeyConfirmation, err = read32(reader); err != nil {
		return nil, family, err
	}
	for index := range out.WithdrawalEncryptedSenderAddress {
		if out.WithdrawalEncryptedSenderAddress[index], err = read32(reader); err != nil {
			return nil, family, err
		}
	}
	if out.RecentPositionFloor, err = read32(reader); err != nil {
		return nil, family, err
	}
	if out.ActionBalanceBlinding, err = readFr32(reader); err != nil {
		return nil, family, err
	}
	if out.NK, err = read32(reader); err != nil {
		return nil, family, err
	}
	if out.CNK, err = read32(reader); err != nil {
		return nil, family, err
	}
	if out.AssetPath, err = readMerklePath(reader); err != nil {
		return nil, family, err
	}
	if out.AssetPosition, err = readU64(reader); err != nil {
		return nil, family, err
	}
	if out.AssetIndexedLeaf, err = readIndexedLeaf(reader); err != nil {
		return nil, family, err
	}
	isRegulated, err := readBool(reader)
	if err != nil {
		return nil, family, err
	}
	out.IsRegulated = isRegulated
	if out.RegulatedPrecision, err = readU8(reader); err != nil {
		return nil, family, err
	}
	if out.UnregulatedPrecision, err = readU8(reader); err != nil {
		return nil, family, err
	}
	if out.RoutingAsOfHeight, err = readU64(reader); err != nil {
		return nil, family, err
	}
	if out.RoutingNonce, err = read32(reader); err != nil {
		return nil, family, err
	}
	if out.SenderCompliancePath, err = readMerklePath(reader); err != nil {
		return nil, family, err
	}
	if out.SenderCompliancePosition, err = readU64(reader); err != nil {
		return nil, family, err
	}
	if out.SenderCapkAffine, err = readPointAffine(reader); err != nil {
		return nil, family, err
	}
	if out.SenderCnkCommitment, err = read32(reader); err != nil {
		return nil, family, err
	}
	if out.SenderStatus, err = read32(reader); err != nil {
		return nil, family, err
	}
	if out.WithdrawalSeed, err = read32(reader); err != nil {
		return nil, family, err
	}
	if out.WithdrawalRandomizer, err = readFr32(reader); err != nil {
		return nil, family, err
	}
	if out.RequiredSpend, err = readShieldedIcs20WithdrawalRequiredSpend(reader); err != nil {
		return nil, family, err
	}
	if out.OptionalSpend, err = readShieldedIcs20WithdrawalOptionalSpend(reader); err != nil {
		return nil, family, err
	}
	if out.ChangeOutput, err = readShieldedIcs20WithdrawalChange(reader); err != nil {
		return nil, family, err
	}
	if out.AKAffine, err = readPointAffine(reader); err != nil {
		return nil, family, err
	}
	if out.AssetIndexedLeafDKPubAffine, err = readPointAffine(reader); err != nil {
		return nil, family, err
	}
	if out.AssetIndexedLeafRingPKAffine, err = readPointAffine(reader); err != nil {
		return nil, family, err
	}
	if out.SenderDiversifiedGenerator, err = readPointAffine(reader); err != nil {
		return nil, family, err
	}
	if reader.Len() != 0 {
		return nil, family, fmt.Errorf("trailing bytes in shielded ICS-20 withdrawal witness: %d", reader.Len())
	}
	return out, family, nil
}

func readShieldedIcs20WithdrawalRequiredSpend(reader *bytes.Reader) (ShieldedIcs20WithdrawalRequiredSpendWitnessBinary, error) {
	var out ShieldedIcs20WithdrawalRequiredSpendWitnessBinary
	var err error
	if out.Nullifier, err = read32(reader); err != nil {
		return out, err
	}
	if out.SpentNoteBlinding, err = read32(reader); err != nil {
		return out, err
	}
	if out.SpentNoteAmount, err = read32(reader); err != nil {
		return out, err
	}
	if out.SpentNoteRecoveryCommitment, err = read32(reader); err != nil {
		return out, err
	}
	if out.StateCommitmentPosition, err = readU64(reader); err != nil {
		return out, err
	}
	if out.StateCommitmentAuthPath, err = readTriplePath(reader); err != nil {
		return out, err
	}
	if out.SpendAuthRandomizer, err = readFr32(reader); err != nil {
		return out, err
	}
	if out.RKAffine, err = readPointAffine(reader); err != nil {
		return out, err
	}
	if out.HistoryRequired, err = readBool(reader); err != nil {
		return out, err
	}
	return out, nil
}

func readShieldedIcs20WithdrawalOptionalSpend(reader *bytes.Reader) (ShieldedIcs20WithdrawalOptionalSpendWitnessBinary, error) {
	required, err := readShieldedIcs20WithdrawalRequiredSpend(reader)
	if err != nil {
		return ShieldedIcs20WithdrawalOptionalSpendWitnessBinary{}, err
	}
	isDummy, err := readBool(reader)
	if err != nil {
		return ShieldedIcs20WithdrawalOptionalSpendWitnessBinary{}, err
	}
	dummyNullifierSeed, err := read32(reader)
	if err != nil {
		return ShieldedIcs20WithdrawalOptionalSpendWitnessBinary{}, err
	}
	return ShieldedIcs20WithdrawalOptionalSpendWitnessBinary{
		ShieldedIcs20WithdrawalRequiredSpendWitnessBinary: required,
		IsDummy:            isDummy,
		DummyNullifierSeed: dummyNullifierSeed,
	}, nil
}

func readShieldedIcs20WithdrawalChange(reader *bytes.Reader) (ShieldedIcs20WithdrawalChangeWitnessBinary, error) {
	var out ShieldedIcs20WithdrawalChangeWitnessBinary
	var err error
	if out.NoteCommitment, err = read32(reader); err != nil {
		return out, err
	}
	if out.RecoveryCommitment, err = read32(reader); err != nil {
		return out, err
	}
	if out.CreatedNoteBlinding, err = read32(reader); err != nil {
		return out, err
	}
	if out.CreatedNoteAmount, err = read32(reader); err != nil {
		return out, err
	}
	if out.RecoveryCapsule, err = readRecoveryCapsule(reader); err != nil {
		return out, err
	}
	return out, nil
}
